### DISCUSSION TITLE
How can I change the colors using the Blazor Boilerplate template?

### DISCUSSION THREAD

**Discussion by chunt001** [ant-design-blazor/ant-design-blazor#979] [Q&A]
I am using the boilerplate template and can get the settings from the configuration panel. When I add them to the appropriate applicationsettings file the color that is set does not seem to apply. The light theme and other settings seem to work but not the primaryColor setting.

Thank you.

**Comment by ElderJames** [comment]
So far, we haven't implement the color palette #290  .
But we can follow the guides from ant design https://ant.design/docs/react/customize-theme to compile a css file with your replaced less variables.

**Reply by chunt001** [reply]
Thank you for the quick reply....what about modifying the footer? I am looking for some of the code to change the links and add custom links but I can't seem to find where these are added on the BasicLayout?

**Comment by ElderJames** [comment]
We do not currently support styling through component parameters. 
You can import your custom CSS into index.html once you've got it.

**Comment by bbo76** [comment]
Hi @ElderJames , could you explain how to set variables because less files are only accessible when the project is published.

**Comment by ElderJames** [comment]
Following is the custom theme variables


```
// Generated by @ant-design/aliyun-theme

@font-size-sm: 12px;
@font-size-base: 12px;
@primary-color: #0070cc;
@border-radius-base: 0;
@border-radius-sm: 0;
@text-color: fade(#000, 65%);
@text-color-secondary: fade(#000, 45%);
@background-color-base: hsv(0, 0, 96%);
@success-color: #1e8e3e;
@error-color: #d93026;
@warning-color: #ffc440;
@info-color: @primary-color;
@danger-color: @error-color;
@processing-color: @primary-color;
@border-color-base: #dedede;
@border-color-split: #dedede;
@outline-width: 0;
@outline-color: #737373;
@input-height-lg: 36px;
@input-height-base: 32px;
@input-height-sm: 24px;
@input-hover-border-color: #737373;
@form-item-margin-bottom: 16px;
@btn-default-bg: #fafafa;
@btn-default-border: #dedede;
@btn-danger-color: #fff;
@btn-danger-bg: @error-color;
@btn-danger-border: @error-color;
@switch-color: @success-color;
@table-header-bg: #fafafa;
@table-row-hover-bg: #fafafa;
@table-padding-vertical: 15px;
@badge-color: @error-color;
@breadcrumb-base-color: @text-color;
@breadcrumb-last-item-color: @text-color-secondary;
@slider-rail-background-color: @background-color-base;
@slider-rail-background-color-hover: #e1e1e1;
@slider-track-background-color: @primary-color;
@slider-track-background-color-hover: @primary-color;
@slider-handle-border-width: 1px;
@slider-handle-color: #dedede;
@slider-handle-color-hover: #dedede;
@slider-handle-color-focus: #dedede;
@slider-handle-color-tooltip-open: #ddd;
@slider-handle-color-focus-shadow: transparent;
@slider-handle-shadow: 1px 1px 4px 0 rgba(0,0,0,.13);
@alert-success-border-color: #dff4e5;
@alert-success-bg-color: #dff4e5;
@alert-info-border-color: #e5f3ff;
@alert-info-bg-color: #e5f3ff;
@alert-error-border-color: #fcebea;
@alert-error-bg-color: #fcebea;
@alert-warning-border-color: #fff7db;
@alert-warning-bg-color: #fff7db;
@radio-button-bg: transparent;
@radio-button-checked-bg: transparent;
@progress-radius: 0;
@tabs-card-gutter: -1px;
@tabs-card-tab-active-border-top: 2px solid @primary-color;
@layout-body-background: #fafafa;

// Patch
@line-height-base: 1.5717;
@input-padding-vertical-lg: 8px;
@input-padding-vertical-base: 6px;

```

**Reply by ElderJames** [reply]
BTW, there a UI for the theme configuration: https://antdtheme.com/

**Reply by bbo76** [reply]
Thank U for the variables, do you have any tutorial to apply this variable to existing css generated in Blazor ? I wanna use web-compiler in visual studio but I don't know how to proceed with Blazor component library in nuget

**Reply by ElderJames** [reply]
You can use gulp. Flowing is our document's gulpfile.js config:

```
var gulp = require("gulp"),
  cleanCss = require("gulp-clean-css"),
  less = require("gulp-less");

var rename = require('gulp-rename');

gulp.task("default", function () {
  return gulp.src('style/index.less')
    .pipe(less({ javascriptEnabled: true }))
    .pipe(cleanCss({ compatibility: 'ie8' }))
    .pipe(rename('docs.css'))
    .pipe(gulp.dest('wwwroot/css'));
});
```

**Reply by EugenePok** [reply]
Hello, I tried to put index.js and index.less generated from https://antdtheme.com/ in both wwwroot folder and styles folder.
![image](https://user-images.githubusercontent.com/43719791/113648536-55f56e80-96bf-11eb-91bf-ee7153cf5715.png)

Then I replace my gulp.js with the one you provided above.

But it turns out the theme still using Ant-Design default theme.
Did I missed out anything? How to make the custom theme work?

**Reply by NPadrutt** [reply]
@ElderJames  I am a bit puzzled. I tried to use gulp and include a Post Build Task. But as far as I can see the path _content/AntDesign/less/ant-design-blazor.less isn't available at that time since it is only created at publish time. Or would I just have to generate a css once and deploy it fix with my application and regenerate it once a new ant design version comes out?

**Reply by ElderJames** [reply]
Please wait for the 0.8 release today, which include theming less.

**Reply by NPadrutt** [reply]
ah thank you for the tip!
I tried to implement it analog to the doc page and your pull request with 0.8, but so far haven't managed it. It seems I haven't figured out how to correctly include the color.less file

I have now those css / less includes:


    <link href="_content/AntDesign/css/ant-design-blazor.css" rel="stylesheet" />
    <link rel="stylesheet/less" type="text/css" href="css/color.less">

    <link href="ApplySupportTool.Client.styles.css" rel="stylesheet">
    <link href="css/app.css" rel="stylesheet">

The color.less is copied from the doc project and in the wwwroot/css folder

![image](https://user-images.githubusercontent.com/1764367/115019456-85735a80-9eb9-11eb-8fa6-2d0fd0e33be8.png)

Plus those scripts:

    <script src="https://cdn.jsdelivr.net/npm/less@2.7.2/dist/less.min.js"></script>

    
    <script src="_content/AntDesign.Charts/g2plot.js"></script>
    <script src="_content/AntDesign.Charts/ant-design-charts-blazor.js"></script>

    
    <script src="_content/AntDesign/js/ant-design-blazor.js"></script>

    <script src="_content/Microsoft.AspNetCore.Components.WebAssembly.Authentication/AuthenticationService.js"></script>
    <script src="_framework/blazor.webassembly.js"></script>


Also in the index I added those functions:

        window.less = {
            async: true
        };

        window.changeColor = function (hex, ref) {
            window.less
                .modifyVars({
                    '@primary-color': hex
                })
        }

And call it in my program.cs on startup:

        private async static Task SetColorAsync(WebAssemblyHost host)
        {
            IJSRuntime? jsInterop = host.Services.GetRequiredService<IJSRuntime>();
            await jsInterop.InvokeVoidAsync("changeColor", "#444277");
        }

Now I get on startup: 
![image](https://user-images.githubusercontent.com/1764367/115019572-a9cf3700-9eb9-11eb-929c-53b6a772fccc.png)

I also tried to change the path in the index.html to the following, with the same error.

    <link rel="stylesheet/less" type="text/css" href="_content/AntDesign/less/style/color/color.less">

Any idea what I'm missing?

**Reply by ElderJames** [reply]
Yes, I also found the `.less` file can't be output by asp.net core... It seems that the less files are not be supported by default.
And I can't find any solution. But it works well with Blazor WebAssembly.

**Reply by NPadrutt** [reply]
My application is Blazor WebAssembly. 
I tried to do it analog to the doc page in the repo. But that one accesses the .less files directly and not after the compilation as far as I saw.

**Reply by ElderJames** [reply]
Could you try to compile your costom css file?
Just need to follow our less file
https://github.com/ant-design-blazor/ant-design-blazor/blob/master/site/AntDesign.Docs/docs.aliyun.less

**Reply by NPadrutt** [reply]
I tried that, but the problem here is, that I can't access it like this AFAIK. The path used (for example`'../../components/ant-design-blazor.aliyun.less'`) is only available if you have those projects in your solution, but not when you use AntBlazor via the NuGet.

I tried before to adjust that for `_content/AntDesign/less/..` but in the end I didn't manage to get this working with gulp. It always was missing one of the files.. 

For now I downloaded the 0.8 release, unzipped it and made a gulp file directly to the `ant-design-blazor.less` there to create a static custom .css of the dark and the light theme and reference that. Not elegant and dynamic changing of the color doesn't work I guess. but it is enough for me at the moment.

**Reply by ElderJames** [reply]
Yes, you are right. The less files was not be put in user's solution. So there have another way is to install the `antd` npm package...
